cabal-install/src/Distribution/Client/ProjectPlanning/Types.hs

```diff
@@ -750,18 +782,39 @@ data BuildStyle =
     -- Typically 'BuildAndInstall' packages will only depend on other
     -- 'BuildAndInstall' style packages and not on 'BuildInplaceOnly' ones.
     --
-  | BuildInplaceOnly
+  | BuildInplaceOnly MemoryOrDisk
+    -- | Built in-memory only using GHC multi-repl, they are not built or installed
+    -- anywhere on disk. BuildInMemory packages can't be depended on by BuildAndInstall nor BuildInplaceOnly packages
+    -- (because they don't exist on disk) but can depend on other BuildStyles.
+
+    -- At the moment BuildInMemory is only used by the 'repl' command.
+--  | BuildInMemory
   deriving (Eq, Show, Generic)
 
-instance Binary BuildStyle
-instance Structured BuildStyle
+data MemoryOrDisk = InMemory | OnDisk  deriving (Eq, Show, Generic)
+
+isInplaceBuildStyle :: BuildStyle -> Bool
+isInplaceBuildStyle (BuildInplaceOnly {}) = True
+isInplaceBuildStyle BuildAndInstall = False
+
+instance Binary MemoryOrDisk
+instance Structured MemoryOrDisk
+
+instance Semigroup MemoryOrDisk where
+  InMemory <> x = x
+  OnDisk <> _   = OnDisk
```

Comment:
This requires a comment in my opinion. I can imagine, this helps ensuring the closure property, but this needs to be documented here, *why* we do it this way.

Reply:
This code has changed now.